tests/test_flows.py

```diff
@@ -2416,6 +2441,159 @@ def my_flow():
         assert my_mock.call_args_list == [call(), call()]
 
 
+class TestFlowHooksOnCancellation:
+    def test_on_cancellation_hooks_run_on_cancelled_state(self):
+        my_mock = MagicMock()
+
+        def cancelled_hook1(flow, flow_run, state):
+            my_mock("cancelled_hook1")
+
+        def cancelled_hook2(flow, flow_run, state):
+            my_mock("cancelled_hook2")
+
+        @flow(on_cancellation=[cancelled_hook1, cancelled_hook2])
+        def my_flow():
+            return State(type=StateType.CANCELLING)
+
+        my_flow._run()
+        assert my_mock.mock_calls == [call("cancelled_hook1"), call("cancelled_hook2")]
+
+    def test_on_cancellation_hooks_are_ignored_if_terminal_state_completed(self):
+        my_mock = MagicMock()
+
+        def cancelled_hook1(flow, flow_run, state):
+            my_mock("cancelled_hook1")
+
+        def cancelled_hook2(flow, flow_run, state):
+            my_mock("cancelled_hook2")
+
+        @flow(on_cancellation=[cancelled_hook1, cancelled_hook2])
+        def my_flow():
+            return State(type=StateType.COMPLETED)
+
+        my_flow._run()
+        assert my_mock.mock_calls == []
+
+    def test_on_cancellation_hooks_are_ignored_if_terminal_state_failed(self):
+        my_mock = MagicMock()
+
+        def cancelled_hook1(flow, flow_run, state):
+            my_mock("cancelled_hook1")
+
+        def cancelled_hook2(flow, flow_run, state):
+            my_mock("cancelled_hook2")
+
+        @flow(on_cancellation=[cancelled_hook1, cancelled_hook2])
+        def my_flow():
+            return State(type=StateType.FAILED)
+
+        my_flow._run()
+        assert my_mock.mock_calls == []
```

Comment:
Should generally use `assert_not_called()`

Reply:
Good `call`!